File: kotlin/src/main/java/ValidParentheses.java
import java.util.Deque;
import java.util.LinkedList;

public class ValidParentheses {
    public boolean isValid(String s) {
        if (s.length() % 2 == 1) return false;
        Deque<Character> stack = new LinkedList<>();
        for (int i =0; i< s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) return false;
                if (c == ')' && stack.pop() != '(') {
                    return false;
                }
                if (c == ']' && stack.pop() != '[') {
                    return false;
                }
                if (c == '}' && stack.pop() != '{') {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        ValidParentheses vp = new ValidParentheses();
        boolean result = vp.isValid("()");
        System.out.println("result=" + result);
    }
}

File: kotlin/src/main/kotlin/ValidParentheses.kt
package org.wyekings

import java.util.Deque
import java.util.LinkedList
import java.util.Stack

class ValidParentheses {
    fun isValid(s: String): Boolean {
        val length = s.length
        if (length % 2 == 1) return false

        val pairs = hashMapOf(
            ')' to '(',
            ']' to '[',
            '}' to '{',
        )

        val stack: Deque<Char> = LinkedList()
        s.forEach { c ->
            if (pairs.containsKey(c)) {
                if (stack.isEmpty() || stack.peek() != pairs[c]) {
                    return false
                }
                stack.pop()
            } else {
                stack.push(c)
            }
        }

        return false
    }
}